core/arch/interp.c

```diff
@@ -5771,6 +5771,172 @@ mangle_x64_ib_in_trace(dcontext_t *dcontext, instrlist_t *trace, instr_t *target
+ * trace_next_exit with tgt_reg directly and remember to restore the value of tgt_reg
+ * in the case when the branch is not taken.
+ *
+ */
+
+static reg_id_t
+check_patched_ibl(dcontext_t *dcontext, instrlist_t *trace, instr_t *targeter,
+                  int *added_size, bool *tgt_in_stolen_reg)
+{
+    instr_t *prev = instr_get_prev(targeter);
+
+    for (prev = instr_get_prev_expanded(dcontext, trace, targeter); prev;
+         prev = instr_get_prev(prev)) {
+
+        instr_t *prev_prev = instr_get_prev(prev);
+        if (prev_prev == NULL)
+            break;
+
+        /* we expect
+         * prev_prev   str IBL_TARGET_REG, TLS_REG2_SLOT
+         * prev        mov IBL_TARGET_REG, jump_target_reg
+         */
+
+        if (instr_get_opcode(prev_prev) == OP_str && instr_get_opcode(prev) == OP_orr &&
+            opnd_get_reg(instr_get_src(prev_prev, 0)) == IBL_TARGET_REG &&
+            opnd_get_base(instr_get_dst(prev_prev, 0)) == dr_reg_stolen &&
+            opnd_get_reg(instr_get_dst(prev, 0)) == IBL_TARGET_REG) {
+
+            reg_id_t jp_tg_reg = opnd_get_reg(instr_get_src(prev, 1));
+
+            instrlist_remove(trace, prev_prev);
+            instr_destroy(dcontext, prev_prev);
+            instrlist_remove(trace, prev);
+            instr_destroy(dcontext, prev);
+
+            LOG(THREAD, LOG_INTERP, 4, "found and removed str/mov\n");
+            *added_size -= 2 * AARCH64_INSTR_SIZE;
+            return jp_tg_reg;
+
+            /* Here we expect
+             * prev_prev  str   IBL_TARGET_REG, TLS_REG2_SLOT
+             * prev       ldr   IBL_TARGET_REG, TLS_REG_STOLEN_SLOT
+             */
+        } else if (instr_get_opcode(prev_prev) == OP_str &&
+                   instr_get_opcode(prev) == OP_ldr &&
+                   opnd_get_reg(instr_get_src(prev_prev, 0)) == IBL_TARGET_REG &&
+                   opnd_get_base(instr_get_src(prev, 0)) == dr_reg_stolen &&
+                   opnd_get_reg(instr_get_dst(prev, 0)) == IBL_TARGET_REG) {
+            *tgt_in_stolen_reg = true;
+            LOG(THREAD, LOG_INTERP, 4, "jump target is in stolen register slot\n");
+            return IBL_TARGET_REG;
+        }
+    }
+    return DR_REG_NULL;
+}
+
+/* For AArch64 we have a special case if we cannot remove all code after
+ * the direct branch, which is mangled by cbz/cbnz stolen register.
+ * For example:
+ *    cbz x28, target
+ * would be mangled (see mangle_cbr_stolen_reg() in aarchxx/mangle.c) into:
+ *    str x0, [x28]
+ *    ldr x0, [x28, #32]
+ *    cbnz x0, fall       <- meta instr, not treated as exit cti
+ *    ldr x0, [x28]
+ *    b target            <- delete after
+ * fall:
+ *    ldr x0, [x28]
+ *    b fall_target
+ *    ...
+ * If we delete all code after "b target", then the "fall" path would
+ * be lost. Therefore we need to append the fall path at the end of
+ * the trace as a fake exit stub. Swapping them might be dangerous since
+ * a stub trace may be created on both paths.
+ */
+
```

Comment:
Please include `TODO i#xxxx` comments (with the real issue number) for the missing pieces: in particular, adding translation support for all the new mangling code.  (Ideally it would be in this PR but for making progress here w/ a new feature in pieces ok to not have all at once so long as what's missing is clearly documented not just in the commit description but in the code too).  I would put such comments both here and in translate.c.

Reply:
I feel this there isn't any translation support needed for the new mangling code. The new mangling code includes:
```
     * Original ibl lookup:
     *      str tgt_reg, TLS_REG2_SLOT
     *      mov tgt_reg, jump_target
     *      b ibl_routine
     * Now we rewrite it into:
     *      str x0, TLS_REG0_SLOT
     *      mov x0, #trace_next_target
     *      eor x0, x0, jump_target
     *      cbnz x0, trace_exit (ibl routine)
     *      ldr x0, TLS_REG0_SLOT
     */
   ```
   and
   ```
    * For example, the code will be fixed to:
 *      eor x0, x0, jump_target
 *      cbnz x0, trace_exit_label
 *      ...
 *      b trace_head
 * trace_exit_label:
 *      ldr x0, TLS_REG0_SLOT
 *      str x2, TLS_REG2_SLOT
 *      mov x2, jump_target
 *      b ibl_routine
   ```
which are just basic spilling of registers and perform some calculation after spilling, I think? What sort of support do you think would be needed for these new mangling instructions?

Reply:
The issue I mentioned previously about `recreate_fragment_ilist` not being able to find an instruction was purely because the new mangled instructions added by `fixup_indirect_trace_exit` is not added when creating the fragment ilist.

Reply:
Please run with -stress_recreate_state and verify there are no translation failures or warnings about unknown mangling sequences.

Reply:
Looks like the "unsupported mangle instr" message is at `DOLOG(2` -- I thought it was more visible esp with -stress_recreate_state.  Please run with `-loglevel 2 -stress_recreate_state` and ensure there are no "unsupported mangle instr" messages in the log files.

Reply:
I think there are already quite a lot of unsupported mangle instr even without trace turned. I was not able to find any issues mentioning this so I opened a new one #5069. Feel free to point out if there are already issues that are talking about this problem.

I will also add a inline comment that references this translation support issue.

Reply:
I added a comment in translate.c referencing issue #5069 but I didn't add any in interp.c because I believe all those mangled instructions added by trace support will be covered by the condition `instr_is_trace_cmp` and never reach the `DOLOG(2`. And again, correct me if I am wrong, but I don't think we have introduced any instructions that require additional support.

Reply:
In general we really prefer linear control flow and a lot of code assumes linear control flow, including translation.  This linear requirement used to be a strong invariant but various issues have required adding limited internal control flow.  This PR is adding internal control flow with different spill/restore logic on different paths, which is going to break translate.c's assumptions on linear control flow.  There's a spill that's not restored until trace_exit_label and the translate code will see "spill x0; restore x0; restore x0" and probably freak out -- when you ran with stress did it assert?  I think it focuses on "our mangling" regions -- I would double-check that the trace_exit_label code is properly marked as "our mangling", and that the stress code walks the list all the way to the end including these trace_exit_label sequences.  This trace_exit_label complexity is going to cause headaches with translation and require special support.  IMHO alternatives should be strongly considered such as the proposal I had earlier.

Reply:
> I think there are already quite a lot of unsupported mangle instr even without trace enabled. I was not able to find any issues mentioning this so I opened a new one #5069. Feel free to point out if there are already issues that are talking about this problem.

This is surprising since I put in work earlier this year to enable -stress_recreate_state, fix a number of translation problems on aarch64, and enable a regression test: #4680.  Taking a look.

Reply:
To make progress let's merge this but please first add a comment in the code and in the commit message about these separate trace_exit_label sequences breaking translation and how either special support needs to be added or (the better choice IMHO: best to avoid internal control flow) they need to be removed and a different scheme used such as the one proposed at https://github.com/DynamoRIO/dynamorio/pull/5045/files/b9ada873ecc0b86128169aae0bd7ad648c0f921b..fc70e80bad8a66b7f417c27673fa5659e96ff1ef#r697566139

Reply:
> the better choice IMHO: best to avoid internal control flow

I agree with you that a linear control flow is preferable, will add a comment in the code for `trace_exit_label

Reply:
> This is surprising since I put in work earlier this year to enable -stress_recreate_state, fix a number of translation problems on aarch64, and enable a regression test: #4680. Taking a look.

I am guessing it's because these instructions are just reported with a log message, which did not catch your attention while you were fixing them?

Reply:
I recall looking at such log messages -- but I think those memories are from rseq translation work #4316 and the #4680 is as you said: focusing on asserts and fatal problems.